yearn/partners/delegated.py

```diff
@@ -0,0 +1,110 @@
+from collections import defaultdict
+from math import ceil
+
+from brownie import chain
+from toolz import last
+from yearn.events import decode_logs, get_logs_asap
+from yearn.networks import Network
+from yearn.utils import contract
+
+YEARN_PARTNER_TRACKER = contract({
+    Network.Mainnet:"0x8ee392a4787397126C163Cb9844d7c447da419D8",
+    Network.Fantom:"0x086865B2983320b36C42E48086DaDc786c9Ac73B",
+}[chain.id])
+
+class AsOfDict(dict):
+    """
+    Allows you to get the closest previous item.
+    
+    >>> AsOfDict({1: 'a', 2: 'b'})[2]
+    'a'
+    """
+
+    def __getitem__(self, key):
+        return super().__getitem__(last(item for item in sorted(self) if item <= key))
+
+def delegated_deposit_totals():
+    delegated_deposits = defaultdict(lambda: defaultdict(lambda: defaultdict(AsOfDict)))
+    for deposit in decode_logs(get_logs_asap(str(YEARN_PARTNER_TRACKER), [YEARN_PARTNER_TRACKER.topics['ReferredBalanceIncreased']])):
+        partnerId, vault, depositor, amount_added, total_deposited = deposit.values()
+
+        vault_deposits = delegated_deposits[vault]
+        depositor_deposits = vault_deposits[depositor]
+        partner_deposits = depositor_deposits[partnerId]
+
+        # Make sure AsOfDict has a start point.
+        if len(partner_deposits) == 0:
+            partner_deposits[0] = 0
```

Comment:
do the other two deposits also need a default value ?
`vault_deposits`, `depositor_deposits`

Reply:
This is resolved with your addition of an __init__ method to AsOfDict